### DISCUSSION TITLE
PaintShop Pro (.psp, .pspimage) file support

### DISCUSSION THREAD

**Discussion by tannerhelland** [tannerhelland/PhotoDemon#347] [Ideas]
Do you use Corel PaintShop Pro (either the modern Corel or the older JASC versions)?  If so, I'd love to get your help.  

I'm working on adding PaintShop Pro file support to PhotoDemon, but the last copy of PSP I owned is from the 1990s (version 4).  I'm assuming the software has changed slightly since then...  😆 

If you have any PSP files on your PC that you don't mind publicly sharing, please attach them to this discussion.  The more, the merrier - I'll be using these files to test PhotoDemon's PSP engine as I make progress on it.

Thank you!

**Comment by Kroc** [comment]
I have many hundreds of PSP images in v6 and v10 formats, it was the main editor I used for my art from around 2000 to 2007. Some of these files are insanely complex (hundreds of layers), although vectors feature predominantly, so I don't know if you'll be able to render them.

**Reply by tannerhelland** [reply]
Alright, I think I'm ready for whatever you're willing to send over.  Feel free to attach a .zip here or just mail 'em directly.  Thank you in advance!

**Comment by tannerhelland** [comment]
Actually, vector layers are one of the things I'm most interested in supporting!

The main reason I've been aggressively optimizing my 2D vector toolkit this month (all the recent pd2D commits) is in preparation for vector layers in PD.  But I thought it'd be worthwhile to investigate PaintShop Pro's support for vector layers first, since they're actually well-documented, unlike Photoshop's.  

So yes, vector layers are high on my to-do list, and PSP files are a much easier place for me to start than something like SVG (which are horrifically complex by comparison).

**Reply by Kroc** [reply]
In that case, I believe I have a file with 1'000 vector layers made up of complex shapes with hundreds of custom gradients throughout :P

**Reply by tannerhelland** [reply]
(I'm gonna regret asking for this, aren't I... 😨 🤣  )

In all seriousness, thank you - complex images are indispensable tools for fuzzing a decoder.

v6 images will also be helpful as the earliest PSP specs I can find online are v7+, so v6 files may expose unexpected changes I'll need to account for.

**Reply by Kroc** [reply]
100% vector! 
![image](https://user-images.githubusercontent.com/142650/103448673-c93cae00-4c94-11eb-8480-77b8a9d1fad0.png)

This is the craziest vector image I have: 
![image](https://user-images.githubusercontent.com/142650/103448703-32242600-4c95-11eb-88c3-de22e71b8b44.png)

**Reply by Kroc** [reply]
Is that several hundred vector trees I see before me??
![image](https://user-images.githubusercontent.com/142650/103448820-bd51eb80-4c96-11eb-9bef-30a70de1e504.png)

**Reply by tannerhelland** [reply]
Well, if trying to make these work in PD doesn't reduce my old VB6 development PC to a heap of smoldering ruins, I suppose nothing will... 🔥 lol

**Reply by Kroc** [reply]
I've sent a few files to you, hopefully they don't fall into the junk pile!

**Reply by tannerhelland** [reply]
> hopefully they don't fall into the junk pile!

They did, alas, but they've been recovered now!  Thank you very much for sending these over.  I'll try to post a quick comment in this thread as I (hopefully) make progress on reading and rendering vector layers.

**Reply by tannerhelland** [reply]
Quick progress update!  Here's a screenshot from the latest PhotoDemon nightly, using the same source PSP file as the screenshot immediately above (c/o @Kroc):

![image](https://user-images.githubusercontent.com/1930029/104645217-05e0b000-566c-11eb-9e2a-a6fa978b8a6c.png)

The trees, mountains (including highlights), and raindrop impacts are all vector objects from the base PSP file.  PhotoDemon can now construct equivalent GDI+ paths, then fill/stroke them very closely to their PSP originals.  Despite the aforementioned "hundreds of vector trees", the file loads in 2-3 seconds on my old VB6 PC, compared to 10+ seconds on a trial version of PSP X6.  (The only thing PSP loads that PhotoDemon does not are a few text layers.  I hope to tackle these next.)

For comparison, here's a screenshot from a debug build, using a red pen to stroke PD's reconstruction of the various bezier paths:

![image](https://user-images.githubusercontent.com/1930029/104645954-0b8ac580-566d-11eb-9fa5-5b858d91ffd9.png)

**Comment by hi5** [comment]
First of all compliments on your program. I had tested it a few years ago but thought it wasn't for me yet - but I had a look at again today and I'm impressed - very fast and smooth program.

As a diehard PaintShop Pro user (still using v8.10 daily) I'm of course interested in the ability to open pspimage files which seems to work well.

The only thing I notice is that vector text is imported as a raster layer, do you think it will be possible to import/convert text as a basic text layer in Photodemon? The font family, size, colors doesn't matter as such because once it is editable text we can then change it in PhotoDemon and save it as PDI file to continue to work on it (basic example below)

That would making switching very easy for me as I find that with each update of Windows PaintShop Pro becomes a bit more unstable and unusable.

[psp810-vector-text.zip](https://github.com/tannerhelland/PhotoDemon/files/8742407/psp810-vector-text.zip)

**Reply by tannerhelland** [reply]
Hello @hi5 .  Thank you for the kind message!

I have happy news to report.  Current PhotoDemon nightly builds *do* import PSP text layers as vector text - but they will be imported as `advanced text layers`.  This is necessary due to the way some PSP text properties are encoded (these properties cannot be reproduced with the default Windows text renderer, which is what `basic text layers` use for rendering).

If you load your attached PSP file into a current nightly build, then switch to the `advanced text tool` and click the `edit text` button in the toolbar, you'll see that the text is fully editable, and PhotoDemon has attempted to replicate the original PSP formatting where it can.  (As you mention it won't be 100% perfect, but it should be close enough that the image still looks OK!)

It also looks like your sample file may have exposed a bug in my current PSP importer, where line breaks in text layers are not recognized correctly.  I will investigate this and report back soon.

Please let me know if you encounter any other issues with PSP import!  I am excited for that feature to receive wider testing.  (During development, I primarily tested PSP 7 - the last version I have a copy of - so testing more v8.10 files will be awesome.)

**Reply by hi5** [reply]
You're correct, it works, I hadn't noticed that it was editable - wonderful. Happy to test further.

**Comment by Kroc** [comment]
Will PhotoDemon support layer groups in the future? One area where .pspimage support falls down is that in PSP8+ blending modes of layers only affect layers [below] within the same layer group and the whole layer group can have its own blending mode which is "global"

**Comment by hi5** [comment]
Just layer groups in general would be nice addition indeed so you can move multiple layers at once vs moving them one by one (perhaps it is already possible and I haven't found how to do it yet)

**Comment by tannerhelland** [comment]
Layer groups and layer masks are both on my TODO list.  Unfortunately, both features are massive projects with two attributes I loath: massive UI requirements, and massive changes needed in PD's internal compositor (the complicated code that renders the main viewport).  These attributes make it very complex to add features like layer groups and masks "in stages" so much like the recent selection tool overhaul, they will need to be slowly worked on in separate branches over a number of months.

I apologize but the ETA for these will not be any time soon 😞

**Reply by hi5** [reply]
No rush, look forward to testing it out when the time comes.
